### ISSUE TITLE
Build Fail

### ISSUE THREAD

**Issue by redknight1056** [sat8/tinkerboard-yocto-qt5-eglfs#2]
I seem to have a build failure for both meta-toolchain-qt5 and core-image-base.

First off is the following item shown below expected at build configuration as it is certainly different to the rockchip build is terms of the 'unknown' stated parts;

`Build Configuration:
BB_VERSION        = "1.35.0"
BUILD_SYS         = "x86_64-linux"
NATIVELSBSTRING   = "universal"
TARGET_SYS        = "arm-rk-linux-gnueabi"
MACHINE           = "tinker-rk3288"
DISTRO            = "rk-none"
DISTRO_VERSION    = "2.4"
TUNE_FEATURES     = "arm armv7ve vfp neon vfpv4 callconvention-hard cortexa17"
TARGET_FPU        = "hard"
meta              
meta-poky         
meta-oe           
meta-multimedia   
meta-networking   
meta-filesystems  
meta-python       
meta-qt5          
meta-rockchip     
meta-rockchip-extra = "<unknown>:<unknown>"`

The actual build issues seems to be around the qtwebkit, I would include the actual log file, but it's 784MB for some reason. I have added a small part of the log would I believe shows the issue.

`EPNS_5EventE+0x550): undefined reference to `WebCore::InspectorInstrumentation::s_frontendCounter'
| collect2: error: ld returned 1 exit status
| Makefile.api:94: recipe for target '../lib/libQt5WebKit.so.5.9.2' failed
| make[2]: *** [../lib/libQt5WebKit.so.5.9.2] Error 1
| make[2]: Leaving directory '/home/ben/tinkerboard-yocto-qt5-eglfs-master/build/tmp/work/cortexa17hf-neon-vfpv4-rk-linux-gnueabi/qtwebkit/5.9.2+gitAUTOINC+97c4a80a12-r0/build/Source'
| Makefile.QtWebKit:46: recipe for target 'sub-api-pri-make_first-ordered' failed
| make[1]: *** [sub-api-pri-make_first-ordered] Error 2
| make[1]: Leaving directory '/home/ben/tinkerboard-yocto-qt5-eglfs-master/build/tmp/work/cortexa17hf-neon-vfpv4-rk-linux-gnueabi/qtwebkit/5.9.2+gitAUTOINC+97c4a80a12-r0/build/Source'
| Makefile:353: recipe for target 'sub-Source-QtWebKit-pro-make_first-ordered' failed
| make: *** [sub-Source-QtWebKit-pro-make_first-ordered] Error 2
| ERROR: oe_runmake failed
| WARNING: exit code 1 from a shell command.
| ERROR: Function failed: do_compile (log file is located at /home/ben/tinkerboard-yocto-qt5-eglfs-master/build/tmp/work/cortexa17hf-neon-vfpv4-rk-linux-gnueabi/qtwebkit/5.9.2+gitAUTOINC+97c4a80a12-r0/temp/log.do_compile.6306)
ERROR: Task (/home/ben/tinkerboard-yocto-qt5-eglfs-master/sources/meta-qt5/recipes-qt/qt5/qtwebkit_git.bb:do_compile) failed with exit code '1'
NOTE: Tasks Summary: Attempted 2691 tasks of which 2268 didn't need to be rerun and 1 failed.

Summary: 1 task failed:
`

**Comment by spedagadi** [comment]
@redknight1056 the log file being 784MB is not sounding good.

Here is a something you can try

      $ repo init -u https://github.com/rockchip-linux/manifests -b yocto
      $ repo sync

Once you have all the yocto files, type
     
      $ . ./setup-environment
You will be prompted to run the command once more, so type 
      
      $ . ./setup-environment

Then, replace contents of **build/local.conf** file of your local directory with the contents of the same [file](https://github.com/sat8/tinkerboard-yocto-qt5-eglfs/blob/master/build/conf/local.conf) present in this repo. Then type 

     $ bitbake core-image-base

The idea here is to try the latest sources from Rockchip than to use mine to check if it makes a difference. (in case things are changing on Rockhchip side, this experiment will definitely help)

**Comment by redknight1056** [comment]
Thank you for the info, I've tried the steps shown above and now get an different error which is:
`ERROR: qt5everywheredemo-1.0-r0 do_patch: Command Error: 'quilt --quiltrc /home/ben/tinkerboard-yocto-qt5-eglfs-master/build/tmp/work/cortexa17hf-neon-vfpv4-rk-linux-gnueabi/qt5everywheredemo/1.0-r0/recipe-sysroot-native/etc/quiltrc push' exited with 0  Output:
Applying patch 0001-show-maximized-window.patch
patching file main.cpp
Hunk #1 FAILED at 9.
1 out of 1 hunk FAILED -- rejects in file main.cpp
Patch 0001-show-maximized-window.patch does not apply (enforce with -f)
ERROR: qt5everywheredemo-1.0-r0 do_patch: Function failed: patch_do_patch
ERROR: Logfile of failure stored in: /home/ben/tinkerboard-yocto-qt5-eglfs-master/build/tmp/work/cortexa17hf-neon-vfpv4-rk-linux-gnueabi/qt5everywheredemo/1.0-r0/temp/log.do_patch.2053
ERROR: Task (/home/ben/tinkerboard-yocto-qt5-eglfs-master/sources/meta-qt5/recipes-qt/examples/qt5everywheredemo_1.0.bb:do_patch) failed with exit code '1'
`

**Comment by spedagadi** [comment]
@redknight1056 I think we need to work backwards from here to get to a normal state by removing problematic packages. For instance **qt5everywheredemo** is causing an issue here, so I suggest to delete the line ** qt5everywheredemo \ ** in build/local.conf in your local directory. Try bitbake again... 

One more thing, supposing you haven't deleted your old(failed) Yocto directories. The must be a **downloads** folder which can simply be copied across to your new Yocto directory. This saves considerable amount of time as bitbake will just need compile rather than to download & compile.

**Comment by redknight1056** [comment]
Good plan.
I'll start the building process now and report any issue's

Certainly bad luck that something had changed  in the time between you making your build image and me trying to do the same.

I wonder if there is a way to package up the toolchain, in a similar way to the build image your provided via google drive.

**Comment by spedagadi** [comment]
>>I wonder if there is a way to package up the toolchain, in a similar way to the build image your provided >>via google drive.

@redknight1056 I can try doing this but you will be needing the sysroot to let the compiler in IDE (Eclipse/QTCreator etc.,) to link against the correct libraries for cross compilation from a Linux 32/64 bit build PC. 

I use Eclipse [Yocto ADT plugin](https://www.yoctoproject.org/tools-resources/projects/eclipse-ide-plug) and always point the sysroot and toolchain within the IDE.

**Comment by redknight1056** [comment]
After another build, there is another issue around the following:
`ERROR: boost-1.64.0-r0 do_compile: Function failed: do_compile (log file is located at `/home/ben/tinkerboard-yocto-qt5-eglfs-master/build/tmp/work/cortexa17hf-neon-vfpv4-rk-linux-gnueabi/boost/1.64.0-r0/temp/log.do_compile.3932)`

For some reason the build sure loves to generate massive error logs, which is not useful for sending to support :/

In terms of adjusting syslinks I wonder if this can be used:
`wget https://raw.githubusercontent.com/riscv/riscv-poky/master/scripts/sysroot-relativelinks.py
chmod +x sysroot-relativelinks.py
./sysroot-relativelinks.py sysroot`

from [http://blog.qt.io/blog/2017/05/03/qt-git-tinkerboard-wayland/](url)

**Comment by redknight1056** [comment]
Hmm interesting thought but this might just work.

You mention about the downloads folder which should contain all the files needed to build the image.
I'm wondering if you could send your download folder with it content's and when I try building straight from your build instructions without rockchip.

I guess there could be a potential problem as I'm not sure how large your download folder is.

I'm thinking that would get me to your build state.

**Comment by spedagadi** [comment]
@redknight1056 The download folder is massive, I am afraid.

I have built the tool chain and will upload it this evening. Also, out of interest, what is your build PC's spec? My recommendation would be to use Ubuntu 16.04 (64 bit), 8GB RAM (processor wont matter much).

**Comment by redknight1056** [comment]
Nice :D 
I'm currently using Debian 9 64bit bit with 4GB of RAM. I have tried using Ubuntu 16.04 but I seem to have more luck with Debian when it comes to building with Yocto. I have used a VM on my main machine but prefer to build via my laptop so it doesn't affect the performance of my design machine.

**Comment by spedagadi** [comment]
@redknight1056 Ok. I think it will certainly be useful to try on ubuntu 64 bit once more, with the commands provided here (fething repo from Rockchip). All my builds are on ubuntu 16.04 64 bit, so it helps to ensure the build environment is identical to increase the chances of reproducing the issues & fixing them.

**Comment by redknight1056** [comment]
@sat8 No problem I'll try again on Ubuntu when I get home. I'll use your build first without re sync to rockchip.

oh just to double check when I build your image I got the following configuration is this the same as you got ?;

Build Configuration: BB_VERSION = "1.35.0" BUILD_SYS = "x86_64-linux" NATIVELSBSTRING = "universal" TARGET_SYS = "arm-rk-linux-gnueabi" MACHINE = "tinker-rk3288" DISTRO = "rk-none" DISTRO_VERSION = "2.4" TUNE_FEATURES = "arm armv7ve vfp neon vfpv4 callconvention-hard cortexa17" TARGET_FPU = "hard" meta meta-poky meta-oe meta-multimedia meta-networking meta-filesystems meta-python meta-qt5 meta-rockchip meta-rockchip-extra = "<unknown>:<unknown>"

**Comment by spedagadi** [comment]
@redknight1056 yes, the config seems identical.

Just FYI, I have built both the image and SDK/toolchain on a new PC with ubuntu 16.04 and the compilation didnot break with the issues similar to your reports. So I definitely think you should use Ubuntu 16.04 64 bit for working with my Yocto repos.

I have also update readme in this repo with downloadable links for OS image and SDK.

**Comment by redknight1056** [comment]
How Interesting, as I've just had a go on my work PC using VirtualBox with Ubuntu 16.04.02.

And I get the same issues :( It's weird that on my home machine Ubuntu loves to restart for no issue while building the image.

Just to make sure everything is the same as yours I did the following on a newly installed Ubuntu 16.04 LTS 64bit:
 `sudo apt-get install gawk wget git-core diffstat unzip texinfo build-essential chrpath socat cpio python python3 pip3 pexpect libsdl1.2-dev xterm make xsltproc docbook-utils fop dblatex xmlto python-git libssl-dev pv`

pip3 and pexcept, are not found so I had to use: 
`sudo apt-get install python3-setuptools`
`sudo easy_install3 pip`
`sudo easy_install3 pexpect`

I then download your files, open the terminal under your folder then type `. ./setup-environment` twice then proceed to use `bitbake core-image-base`

Is this the same step's you had taken from a blank Ubuntu OS.
 
Thank you for the SDK I'll check it out when I finish work.

**Comment by spedagadi** [comment]
>>Ubuntu loves to restart for no issue while building the image.
@redknight1056 You can play around with the following values in local.conf

BB_NUMBER_THREADS = '4'
PARALLEL_MAKE = '-j 6'

Initially, these were 8 & 12 respectively and my ubuntu box's CPU spiked constantly while building QT packages with an eventual restart later on. I had to reduce these threads to let the CPU breathe and it was ok later on. At your end, these may not be best and require further tuning.

**Comment by redknight1056** [comment]
After getting over the crashing Ubuntu issue with a slight change to 'BB_NUMBER_THREADS' I'm still getting the same issue seen above with qtbase. This was done on a clean version of Ubuntu with the 'downloads' folder being empty. So this was a 100% clean build.

Apart from my install programs shown above, is there anything else you installed?  I'm starting to think this would be the case. But I really can't shake the feeling that something is weird with your pragram, as I can fully build the normal rockchip yocto image but not yours. But I will continue my research.

Also when it comes to the included SDK there seem to be issues with internal Qt linking, meaning program file .pro will not be parsed into the overall project. I did try for reference the SDK which I built from the rockchip which is based from qt 5.8.0 instead of 5.9.2 and that seems to work fine. 

Thank you for your time.

**Comment by redknight1056** [comment]
Success :D  It was indeed an issue with the additional packages that are to be used by Yocto.

I found the following worked well:
-Fully update Ubuntu
-Then install the following:
`sudo apt-get install gawk wget git-core diffstat unzip texinfo gcc-multilib \
     build-essential chrpath socat cpio python python3 python3-pip python3-pexpect \
     xz-utils debianutils iputils-ping libsdl1.2-dev xterm pv` 
`sudo apt-get install make xsltproc docbook-utils fop dblatex xmlto python-git`

-I've noticed that you will often get a “fatal error: openssl/opensslv.h" I used the following to fix it:
`sudo apt-get install python-pip python-dev libffi-dev libssl-dev libxml2-dev libxslt1-dev libjpeg8-dev zlib1g-dev`
`sudo apt-get install python-configargparse python3-configargparse`

Then it come's to using the ./flashtool I would use this:  
`flash_sdcard() {
	sudo dd if=${IMAGE} of=${DEVICE} seek=${SEEK} conv=notrunc
	pv -tpreb ${IMAGE} | sudo dd of=${DEVICE} seek=${SEEK} conv=notrunc
	sync
 }`
As the current version on this build can results in 'input/output' issues when writing to a SD card, this fix has been approved by the rockchip creator.

Thank you for your hard work and help.
